### ISSUE TITLE
Make it easier to pin a version to a specific tag

### ISSUE THREAD

**Issue by sterlinghughes** [apache/mynewt-newt#157]
I'm trying to understand how the newt tool manages repository versions.

Here's my situation -- I don't want to depend on github/master, yet I don't
want to introduce gratuitous incompatibility.

So I would like to redefine the apache-mynewt-core repository to be my own
repository (either my own fork on github, or perhaps an internal github
server). That's an easy change to make in project.yml.  No problem yet.

On my own fork, I like to keep master to be exactly the same as the
upstream master, although it may lag behind as I don't need to run a pull
constantly.   So I'm keeping my own changes in a development branch.

I'd like to specify this in project.yml, but the 'vers' field has some
strict guidelines about what I can use, and it seems that I can only use
master branch, so "0-dev" is "0.0.0" is "master" according to
repository.yml.

I want to have everything I'm building in source control, for my Continuous
Integration engine if nothing else.   At the moment, I have a patch to the
BAS service in mynewt-nimble that I wish to commit.  I believe that it will
be in the official repository "soon" (for some definition of soon), but
that doesn't help the fact that I would like to check in (to my local
repositories my application that depends on this fix, which then entails
that I want to check in the fix too).

From what I'm reading so far, it would be nice if the newt tool were a
little less picky about the format of the 'vers' tag, or had an alternate
tag that would let me bypass 'vers' altogether and let be specify a branch
name directly (which would also alleviate the need to special case "0.0.0"
to being "master").

Or is there a capability to do what I'm looking for that I just can't find
(I don't speak fluent go, so perusing the source hasn't shown me anything
yet).

I agree - more flexibility here would be an improvement.  I think you
have a good understanding of the system.  Here is my summary of how it
currently works:

    * The `project.yml` file consists of "repo specifiers."
    * A repo specifier contains a single "version string."
    * A version string has one of the following forms:
        * Normalized:  "#.#.#"
        * Floating:    "#[.#]-<stability>"

      where <stability> is "dev", "latest", etc.

    Each Mynewt repo contains a `repository.yml` file in its master
    branch.  The `repository.yml` file:
        * Maps floating version strings to normalized version strings.
        * Finally, maps normalized version strings to git commits
          (branch, tag, or hash).

I think this system works well for users who want to use official Mynewt
releases.  However, it may not be the best for more adventurous users.

To solve your problem, I think you'll need to modify your repo's
`repository.yml` file such that "0.0.0" points to the specific commit
hash that you want to pin yourself to.  For example:

    repo.versions:
        "0.0.0": "815254f5166ef3954b214efdd37549814521c9d6"

For the future, I suggest we make the following changes to newt:

1. Allow a repo specifier (in `project.yml`) to directly specify a git
commit (branch, tag, or hash).

2. Allow a `repository.yml` file to map a floating version number
directly to a git commit, rather than requiring an intermediate
normalized version.

Chris

**Comment by sterlinghughes** [comment]
cc @davidzuhn, @ccollins476ad

**Comment by simonratner** [comment]
Edit: Never mind, I see you want to specify this somewhere other than `master`. This makes more sense now. 

Fwiw, I have moved away from using the newt tool to manage the repo dependencies, instead just tracking the required commits as git submodules. Fewer surprises in the CI pipeline, especially around upgrade time; I know exactly what version of the repo I will be getting.

**Comment by ccollins476ad** [comment]
Unfortunately, this isn't as straightforward as I thought it would be.  Consider the following `project.yml` excerpt:

```
repository.apache-mynewt-core:
    type: 'git'
    vers: 'efe8d5f6bc0719b5a1cbb1d3537edafe2bd1c854'
    url: 'git@github.com:apache/mynewt-core.git'
```

The intent is for newt to checkout the `efe8d5` commit of the `mynewt-core` repo.  The problem with this approach is that an arbitrary commit like the one above doesn't correspond to a specific repo version.  Newt needs to know which version of each repo is installed for a number of reasons:
* Calculate repo dependencies
* Determine if the repo version is compatible with the version of newt being used.

Currently, newt looks at the versions specified in `project.yml` to determine which repo versions are installed.  If the `project.yml` file doesn't specify actual versions, there is a problem.

It is quite a big change, but I think the correct solution is for each repo to indicate its current version in a file (`version.yml`?).  The contents of this file would change as new versions are released.  Using this scheme, newt could determine which version a particular commit corresponds to by reading the version file after performing a `git checkout`.

**Comment by sterlinghughes** [comment]
isn't this what the project.state file is for?   it's meant to be a resolved version of the project.yml file representing the actual currently installed version of each repo.

**Comment by davidzuhn** [comment]
I see there being two issues here.   One is determining which source I wish to check out of github.   That can be identified by a commit, tag or branch.   The second is determining some semantic identification as to what's in that source, and that's where the project.yml file comes into play.   I could check something out from git that's entirely not a Newt project, or the wrong versions for dependencies, etc.   The checkout would succeed, but the semantic analysis would fail.    

But these two items ought to be separated.   If project.yml always has to come from master, then changes to other branches or tags also need to be reflected in master, requiring additional work.  And that's work that's incompatible with at least one common workflow when dealing with private copies of upstream repositories.   

I wouldn't mind at all keeping project.yml up to date (for my purposes), but only in the branches where I'm doing my own work.   Branches that are mirrored from elsewhere should remain unmodified.

I'm not keen on having to keep updating my application with a new commit id every time I make a change in that upstream repo.   As things stand today, I don't change '0-dev' every time that master updates -- it follows the branch.   I'd really like to keep that capability.  I just want to be able to specify my own branches.

**Comment by ccollins476ad** [comment]
@sterlinghughes - Yes, you're right, `project.state` keeps track of the currently-installed repo versions. However, this information is ultimately derived from the version specified in `project.yml`.  I glossed over some details in my previous comment because I felt like it was already getting too wordy :).

Anyway, I don't believe `project.state` helps us here.  For example, if the user wants to pin `apache-mynewt-core` to a particular commit which does not necessarily map to a release tag, newt has no idea which repo version that commit corresponds to.  Therefore, newt cannot populate `project.state` with a version number for `apache-mynewt-core`.

**Comment by ccollins476ad** [comment]
@davidzuhn - I think there is some confusion here, but I'm not sure if it is on my side, yours, or both!  Hopefully we can resolve the confusion.

> I see there being two issues here. One is determining which source I wish to check out of github. That can be identified by a commit, tag or branch. The second is determining some semantic identification as to what's in that source, and that's where the project.yml file comes into play. I could check something out from git that's entirely not a Newt project, or the wrong versions for dependencies, etc. The checkout would succeed, but the semantic analysis would fail.

OK

> But these two items ought to be separated. If project.yml always has to come from master, then changes to other branches or tags also need to be reflected in master, requiring additional work. And that's work that's incompatible with at least one common workflow when dealing with private copies of upstream repositories.

To clarify, it is `repository.yml` that needs to come from master.  `project.yml` is particular to the user's workspace and is not associated with any Mynewt repos.  `project.yml` specifies the set of Mynewt repos, with version numbers, that the project requires.  A `repository.yml` file, on the other hand, is specific to an upstream repo.  Each Mynewt repo contains a `repository.yml` file in its master branch.  `repository.yml`, among other things, maps version numbers of the repo to git commit IDs.

Currently, a repo doesn't "know" which version it is.  Newt knows the repo version because it remembers which version the user requested in his `project.yml` file.  My thought is to incorporate version information into a Mynewt repo such that its current version can be inferred by looking at what has been checked out.

> I wouldn't mind at all keeping project.yml up to date (for my purposes), but only in the branches where I'm doing my own work. Branches that are mirrored from elsewhere should remain unmodified.

> I'm not keen on having to keep updating my application with a new commit id every time I make a change in that upstream repo. As things stand today, I don't change '0-dev' every time that master updates -- it follows the branch. I'd really like to keep that capability. I just want to be able to specify my own branches.

You wouldn't need to change your application each time an upstream repo is updated.  If the upstream repo's `repository.yml` file maps a version number to the git commit, tag, or branch that you wish to use, then you can just specify that version number in your `project.yml` file.  Typically, `0-dev` maps to `0.0.0`, which maps to `master`.  If you always want to use the latest master, then you can continue specifying `0-dev` in your `project.yml` file.  What this github issue concerns (if I understand it) is the ability to use a commit of an upstream repo which *doesn't* map to a version number.

Does that resolve anything?  I think the confusion might just be the `project.yml` / `repository.yml` mixup.  If it seems I'm still missing something, could you please try to explain again? :)

**Comment by davidzuhn** [comment]
> To clarify, it is repository.yml that needs to come from master.

This is the root of my problem.   if I could specify (in the pkg.yml file for the application) a branch of the repository that I want to be checked out initially and where the repository.yml file should be scanned from, I think my use case would be met.

So to restate what I want to do (and I'll be the first to believe that there's a different or better way to do things):

I am a big fan of Continuous Integration.   To me, CI builds *have* to come from clean, newly checked out repositories.   Which means that I cannot use upstream/master directly, since it's not updated on the same pace that I can do (case in point: the changes you suggested for the BAS service, which I want to use but aren't committed upstream yet).   So I go and update the apps pkg.yml to point to my fork of mynewt-{core,nimble,etc}.   I could maintain my local changes in master on that fork, but my github use patterns say that I should keep my changes in a different branch that's unique to me.   So I work on the 'mystuff' branch.   I will pull from upstream/master on my pace, or add my own changes, etc.  I can have repository.yml in the mystuff branch point to the right values that I want for "0.0.0" or "myownrelease-1.2" or "bob".    

With this, my app still references "@apache-mynewt-core" with version "0-dev", but with my local copy.   

The problem is that newt only reads repository.yml from master.   Since I want to keep master updated with upstream/master, without any local changes, my own repository.yml (in "mystuff") is never used.

If I could put one more field in app/pkg.yml, and use that argument as a "-b <argument>" option to "git clone", then I think the rest of the processing would do the expected thing (since the values in repository.yml in the branch that I want is parsed, and further branch/tag/commit processing can occur).

**Comment by ccollins476ad** [comment]
@davidzuhn - OK, I think I finally see what you mean.  That does seem like an important use case, and one I hadn't considered.

I believe my original idea would address your need.  I was considering a different use case, though, so it may not be ideal.  My proposed solution would work as follows:

* mynewt-core's `master` remains identical to your fork's `master`.

* Instead of pointing to a version of the official mynewt-core repo, your `project.yml` file points to a commit, branch, or tag on your fork:

```
repository.apache-mynewt-core:
    type: git
    vers: myownrelease-1.2
    url: 'git@github.com/davidsfork/mynewt-core.git'
```

*Note:* `project.yml` is not part of `mynewt-core` or your fork; it is local to your Mynewt workspace.

`myownrelease-1.2` is a branch, tag, or commit in your fork.  We would probably want some special notation to distinguish git commits from version numbers here, but let's ignore that for now.

Each package in the project still depends on packages in `apache-mynewt-core`, but because of the customized entry in your `project.yml` file, newt translates the dependencies to use your special branch instead of the official repo.

I do think that your idea--specifying a branch in the app's `pkg.yml` file--would work, but I think we would have difficulty fitting that behavior into newt.  By the time newt parses the `pkg.yml` files, it has already determined which branches of each repo it should be using.  As newt is currently implemented, it is the `project.yml` file which determines which branches / tags / commits of each repo are used by the project.

**Comment by davidzuhn** [comment]
Yes, you're right.   It should be project.yml and not pkg.yml.    (Too many YML files floating around and I'm still getting up to speed on this.).

**Comment by jacobrosenthal** [comment]
Any status on this? Ive been following silently, but wanted to signal my
interest.

I think Im in the same boat. Ive got a build system running but were on
0-dev which isnt ideal. In the future Id presumably happily pin to a mynewt
tag but were moving so fast and tagging so slowly thats not working. I
could check in repos but that has been problematic with submodules and non
technical staff.

The proposed solution of pinning to branches or tags would work well for
me.
